src/serde.rs

```diff
@@ -0,0 +1,87 @@
+use serde::{ser::Serialize, Deserialize};
+
+use crate::{NodeId, NodeRef, Tree};
+
+#[derive(Debug)]
+struct SerNode<'a, T> {
+    value: &'a T,
+    children: Vec<SerNode<'a, T>>,
```

Comment:
This might be premature optimization and is possibly better served via a follow-up MR, but this could become `&'b mut Vec<SerNode<'a, T>>` and borrow from a single buffer provided by `Tree::serialize`. The buffer would then be extended before recursing one level down and truncated after returned (storing the number of children in a separate field). This would avoid/amortize/reuse the many small temporary allocations inherent in the current approach. 

(I do agree that we need to materialize the children as formats like bincode cannot handle Serde sequences without a fixed length known upfront.)

Reply:
Or maybe also something like `&[SerNode<'a, T>]`, I see the benefit of such an idea, but I also agree that we could work in it later, maybe even in a different PR

Reply:
I think we should definitely address this in a future PR